src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -222,14 +252,133 @@ public static String encode(String s, String enc)
+                                i++;
+                                continue;
+                            }
+                        }
+                        surrogateError = true;
+                        ut8Length += 3;
+                    } else {
+                        // 3 bytes
+                        ut8Length += 6;
+                    }
+                }
+            }
+        }
+
+        if (needEncodingCount == 0 && spaceCount == 0) {
+            // not need change
+            return s;
+        }
+
+        if (utf8 & !surrogateError) {
+            return encodeUTF8(s, needEncodingCount, ut8Length);
+        }
+
+        return encodeSlow(s, charset);
+    }
+
+    private static String encodeUTF8(String s, int needEncodingCount, int ut8Length) {
+        byte[] buf = new byte[s.length() + needEncodingCount * 2 + ut8Length];
+
+        int off = 0;
+        for (int i = 0, length = s.length(); i < length; ++i) {
+            char c = s.charAt(i);
+            if (c < 0x80) {
+                if (dotNeedEncoding(c)) {
+                    buf[off++] = (byte) (c == ' ' ? '+' : c);
+                } else {
+                    putEncoded(buf, off, c);
+                    off += 3;
+                }
+            } else if (c < 0x800) {
+                // 2 bytes, 11 bits
+                utf8Put2(buf, off, c);
+                off += 6;
+            } else if (Character.isSurrogate(c)) {
+                // Have a surrogate pair and not handle surrogate error
+                char d = s.charAt(i + 1);
+                utf8PutSurrogatePair(
+                        buf,
+                        off,
+                        Character.toCodePoint(c, d));
+                off += 12;
+                i++;
+            } else {
+                // 3 bytes
+                utf8Put3(buf, off, c);
+                off += 9;
+            }
+        }
+
+        try {
+            return jla.newStringNoRepl(buf, StandardCharsets.ISO_8859_1);
+        } catch (CharacterCodingException cce) {
+            throw new AssertionError(cce);
+        }
+    }
+
+    private static void utf8PutSurrogatePair(byte[] buf, int off, int uc) {
+        putEncoded(buf, off, 0xf0 | (uc >> 18));
+        putEncoded(buf, off + 3, 0x80 | ((uc >> 12) & 0x3f));
+        putEncoded(buf, off + 6, 0x80 | ((uc >> 6) & 0x3f));
+        putEncoded(buf, off + 9, 0x80 | (uc & 0x3f));
+    }
+
+    private static void utf8Put2(byte[] buf, int off, int c) {
+        putEncoded(buf, off, 0xc0 | (c >> 6));
+        putEncoded(buf, off + 3, 0x80 | (c & 0x3f));
+    }
+
+    private static void utf8Put3(byte[] buf, int off, int c) {
+        putEncoded(buf, off, 0xe0 | (c >> 12));
+        putEncoded(buf, off + 3, 0x80 | ((c >> 6) & 0x3f));
+        putEncoded(buf, off + 6, 0x80 | (c & 0x3f));
+    }
+
+    private static void putEncoded(byte[] buf, int off, int c) {
+        buf[off] = '%';
+        int n0 = (c >> 4) & 0xf;
```

Comment:
Can we use the fast hex to latin string via ByteArray (big endian here) thing you added in other patches here, or are they not available yet?

Reply:
The micro benchmark using the table lookup method here will have better numbers, but it may not be effective in actual business scenarios because it may not be called multiple times.